[WIP] Scattered Bid Adapter: New Adapter

almendar opened the pull request:
<!--
Thank you for your pull request! 

Please title your pull request like this: 'Module: Change', eg 'Fraggles Bid Adapter: support fragglerock'

Please make sure this PR is scoped to one change or you may be asked to resubmit. 
 
Please make sure any added or changed code includes tests with greater than 80% code coverage. 

See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.

For any user facing change, submit a link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 

https://github.com/prebid/prebid.github.io/pull/4181

- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change
New adapter for Scattered DSP

<!-- For new bidder adapters, please provide the following
- contact email of the adapter’s maintainer
prebid@scattered.pl

- test parameters for validating bids:
```js
  {
                    bidder: "scattered",
                    params: {
                        bidderDomain: "prebid-test.scattered.eu/bid",
                        test: 0
                    }
}
```

almendar commented:
Hi!
@dgirardi is there anything I can do to make this PR rolling? 😃

dgirardi commented on modules/scatteredBidAdapter.js:
```diff
@@ -0,0 +1,175 @@
+    const ortb2Obj = bidderRequest.ortb2 || {};
+    let { user } = ortb2Obj
+
+    let site = getConfig('site') || {};
```
For a new adapter I believe it's better to not do this, since this way of setting FPD is deprecated - and incomplete (the same should be done for `app` and `device`). We have [this issue](https://github.com/prebid/Prebid.js/issues/9335) open to solve this problem in core, after which adapters would see everything directly under `bidderRequest.ortb2`.

dgirardi commented on modules/scatteredBidAdapter.js:
```diff
@@ -0,0 +1,175 @@
+            cpm: bidResponse.price,
+            creativeId: bidResponse.crid,
+            ttl: 360,
+            netRevenue: bidRequest.netRevenue === 'net',
```
I think `result.netRevenue` would always be false in practice - since I don't know of any situation where `bidRequest.netRevenue` would be anything other than undefined. Could you explain the reasoning - who are you expecting to fill it out?

dgirardi commented on modules/scatteredBidAdapter.js:
```diff
@@ -0,0 +1,175 @@
+    // site.page    URL of the page where the impression will be shown.
+    // site.domain  Domain of the site (e.g., “mysite.foo.com”).
+    if (!site.page && bidderRequest.refererInfo) {
+      site.page = bidderRequest.refererInfo.referer;
```
Since version 7 `refererInfo` has `page`, `ref` and `domain` instead of `referer` - see [this PR](https://github.com/prebid/Prebid.js/pull/8450).

dgirardi commented on modules/scatteredBidAdapter.js:
```diff
@@ -0,0 +1,175 @@
+
+    const imps = buildImps(validBidRequests)
+
+    const tid = getKeyOnAny(validBidRequests, 'transactionId');
```
source.tid should be the auctionId, and kept separate from imp transaction ids - see [this issue](https://github.com/prebid/Prebid.js/issues/8573)

dgirardi reviewed (commented):
I apologize for being so late - I completely missed this until you tagged me.

For new adapters that talk to ORTB backends we recommend using our [ORTB conversion library](https://github.com/prebid/Prebid.js/blob/master/libraries/ortbConverter/README.md); besides cutting down on code size, it would give you a free implementation of new Prebid features as they come out.

Since this adapter is so simple, I took the liberty of refactoring it to use that library as an extra incentive - [here's the source tree](https://github.com/dgirardi/Prebid.js/tree/scattered-patch) and the [diff](https://github.com/dgirardi/Prebid.js/commit/c1b5bbdf72eff220d2fbb5be802264d3542d5e64).

I you choose to take in my patch, the library logic invalidates all of my comments below except for the point about  `netRevenue`, where I kept the same behavior which I am not sure makes sense.

almendar commented:
Sure I'll update it. Sorry for delay I was dragged by other task from this. I'll try to fix this soon!

patmmccann commented:
@almendar we're very much looking forward to merging the adapter with @dgirardi 's suggestions

Do you have any concerns if upon merge we were to promote this js as an example of the new 'model' of publishing a concise bid adapter?

almendar commented:
Sorry for being silent!
@patmmccann No, absolutley I have nothing against promoting it :) 

@dgirardi I've applies your patch. Thank you very much for going extra mile and creating it for me :)

I think its ready for second round. In the mean time we've got our own GLV ID so I've added it to the spec. Also updated the sibling docs PR https://github.com/prebid/prebid.github.io/pull/4181

dgirardi commented on modules/scatteredBidAdapter.js:
```diff
@@ -0,0 +1,97 @@
+
+const BIDDER_CODE = 'scattered';
+const GVLID = 1179;
+export function buildImps(validBidRequests) {
```
I think I forgot to remove this in my patch, it doesn't look to be used by anything now.

almendar replied:
Done!

dgirardi commented on modules/scatteredBidAdapter.js:
```diff
@@ -0,0 +1,97 @@
+    mediaType: BANNER,
+    ttl: 360
+  },
+  bidResponse(buildBidResponse, bid, context) {
```
This method is no longer necessary (since it's not overriding anything), but you do need to set `netRevenue` somewhere; if it's always fixed at true or false you can set that through `context: {netRevenue: [true/false], ...}`

almendar replied:
Both done!

almendar commented:
@dgirardi hopefully this time I've got it right :)  Thanks for the review!
3rd round?

dgirardi reviewed (approved):
LGTM, but beware that if I try to use the configuration in the md file, I get no bids - is that expected?

almendar commented:
Yes, this is just an example. We aim for federation of bidders, not all being public.
